# Introductory Econometrics: A modern approach
---

# Chapter 10: Basic Regression Analysis with Time Series Data

> Book: 

    - WOOLDRIDGE, J. M.. Introdução à econometria: uma abordagem moderna. 3a ed. São Paulo: Pioneira Thomson Learning, 2006.

    - WOOLDRIDGE, Jeffrey M. Introductory econometrics: A modern approach. Cengage learning, 2015.

> LINK to codes correction:

http://www.upfie.net/downloads17.html

In [1]:
import wooldridge as woo
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# Example 10.1: Static Phillips Curve

In [2]:
phillips = woo.data('phillips')
phillips_96 = phillips[phillips['year'] <= 1996]
phillips_96

year  unem   inf  inf_1  unem_1  cinf     cunem
0   1948   3.8   8.1    NaN     NaN   NaN       NaN
1   1949   5.9  -1.2    8.1     3.8  -9.3  2.100000
2   1950   5.3   1.3   -1.2     5.9   2.5 -0.600000
3   1951   3.3   7.9    1.3     5.3   6.6 -2.000000
4   1952   3.0   1.9    7.9     3.3  -6.0 -0.300000
5   1953   2.9   0.8    1.9     3.0  -1.1 -0.100000
6   1954   5.5   0.7    0.8     2.9  -0.1  2.600000
7   1955   4.4  -0.4    0.7     5.5  -1.1 -1.100000
8   1956   4.1   1.5   -0.4     4.4   1.9 -0.300000
9   1957   4.3   3.3    1.5     4.1   1.8  0.200000
10  1958   6.8   2.8    3.3     4.3  -0.5  2.500000
11  1959   5.5   0.7    2.8     6.8  -2.1 -1.300000
12  1960   5.5   1.7    0.7     5.5   1.0  0.000000
13  1961   6.7   1.0    1.7     5.5  -0.7  1.200000
14  1962   5.5   1.0    1.0     6.7   0.0 -1.200000
15  1963   5.7   1.3    1.0     5.5   0.3  0.200000
16  1964   5.2   1.3    1.3     5.7   0.0 -0.500000
17  1965   4.5   1.6    1.3     5.2   0.3 -0.700000
18  1966   3.8   2.9    1.6     4.5   1.3 -0.700000
19  1967   3.8   3.1    2.9     3.8   0.2  0.000000
20  1968   3.6   4.2    3.1     3.8   1.1 -0.200000
21  1969   3.5   5.5    4.2     3.6   1.3 -0.100000
22  1970   4.9   5.7    5.5     3.5   0.2  1.400000
23  1971   5.9   4.4    5.7     4.9  -1.3  1.000000
24  1972   5.6   3.2    4.4     5.9  -1.2 -0.300000
25  1973   4.9   6.2    3.2     5.6   3.0 -0.700000
26  1974   5.6  11.0    6.2     4.9   4.8  0.700000
27  1975   8.5   9.1   11.0     5.6  -1.9  2.900000
28  1976   7.7   5.8    9.1     8.5  -3.3 -0.800000
29  1977   7.1   6.5    5.8     7.7   0.7 -0.600000
30  1978   6.1   7.6    6.5     7.1   1.1 -1.000000
31  1979   5.8  11.3    7.6     6.1   3.7 -0.300000
32  1980   7.1  13.5   11.3     5.8   2.2  1.300000
33  1981   7.6  10.3   13.5     7.1  -3.2  0.500000
34  1982   9.7   6.2   10.3     7.6  -4.1  2.100000
35  1983   9.6   3.2    6.2     9.7  -3.0 -0.099999
36  1984   7.5   4.3    3.2     9.6   1.1 -2.100000
37  1985   7.2   3.6    4.3     7.5  -0.7 -0.300000
38  1986   7.0   1.9    3.6     7.2  -1.7 -0.200000
39  1987   6.2   3.6    1.9     7.0   1.7 -0.800000
40  1988   5.5   4.1    3.6     6.2   0.5 -0.700000
41  1989   5.3   4.8    4.1     5.5   0.7 -0.200000
42  1990   5.6   5.4    4.8     5.3   0.6  0.300000
43  1991   6.8   4.2    5.4     5.6  -1.2  1.200000
44  1992   7.5   3.0    4.2     6.8  -1.2  0.700000
45  1993   6.9   3.0    3.0     7.5   0.0 -0.600000
46  1994   6.1   2.6    3.0     6.9  -0.4 -0.800000
47  1995   5.6   2.8    2.6     6.1   0.2 -0.500000
48  1996   5.4   3.0    2.8     5.6   0.2 -0.200000

In [3]:
phillips_96.dtypes

year        int64
unem      float64
inf       float64
inf_1     float64
unem_1    float64
cinf      float64
cunem     float64
dtype: object

In [4]:
# linear regression of static model (Q function avoids conflicts with keywords):

phillips_96_regOLS = smf.ols('Q("inf") ~ Q("unem")', data = phillips_96)
results_phillips_96_regOLS = phillips_96_regOLS.fit()
print(results_phillips_96_regOLS.summary())

                            OLS Regression Results                            
Dep. Variable:               Q("inf")   R-squared:                       0.053
Model:                            OLS   Adj. R-squared:                  0.033
Method:                 Least Squares   F-statistic:                     2.616
Date:                Wed, 27 Jul 2022   Prob (F-statistic):              0.112
Time:                        15:36:03   Log-Likelihood:                -124.43
No. Observations:                  49   AIC:                             252.9
Df Residuals:                      47   BIC:                             256.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.4236      1.719      0.828      0.4

# Example 10.2: Effects of Inflation and Deficits on Interest Rates

In [5]:
intdef = woo.data('intdef')
intdef

year     i3   inf        rec        out       def   i3_1  inf_1     def_1  \
0   1948   1.04   8.1  16.200001  11.600000 -4.600000    NaN    NaN       NaN   
1   1949   1.10  -1.2  14.500000  14.300000 -0.200000   1.04    8.1 -4.600000   
2   1950   1.22   1.3  14.400000  15.600000  1.200001   1.10   -1.2 -0.200000   
3   1951   1.55   7.9  16.100000  14.200000 -1.900001   1.22    1.3  1.200001   
4   1952   1.77   1.9  19.000000  19.400000  0.400000   1.55    7.9 -1.900001   
5   1953   1.93   0.8  18.700001  20.400000  1.699999   1.77    1.9  0.400000   
6   1954   0.95   0.7  18.500000  18.799999  0.299999   1.93    0.8  1.699999   
7   1955   1.75  -0.4  16.500000  17.299999  0.799999   0.95    0.7  0.299999   
8   1956   2.66   1.5  17.500000  16.500000 -1.000000   1.75   -0.4  0.799999   
9   1957   3.27   3.3  17.700001  17.000000 -0.700001   2.66    1.5 -1.000000   
10  1958   1.84   2.8  17.299999  17.900000  0.600000   3.27    3.3 -0.700001   
11  1959   3.41   0.7  16.200001  18.799999  2.599998   1.84    2.8  0.600000   
12  1960   2.93   1.7  17.799999  17.799999  0.000000   3.41    0.7  2.599998   
13  1961   2.38   1.0  17.799999  18.400000  0.600000   2.93    1.7  0.000000   
14  1962   2.78   1.0  17.600000  18.799999  1.199999   2.38    1.0  0.600000   
15  1963   3.16   1.3  17.799999  18.600000  0.800001   2.78    1.0  1.199999   
16  1964   3.55   1.3  17.600000  18.500000  0.900000   3.16    1.3  0.800001   
17  1965   3.95   1.6  17.000000  17.200001  0.200001   3.55    1.3  0.900000   
18  1966   4.88   2.9  17.299999  17.799999  0.500000   3.95    1.6  0.200001   
19  1967   4.32   3.1  18.400000  19.400000  1.000000   4.88    2.9  0.500000   
20  1968   5.34   4.2  17.600000  20.500000  2.900000   4.32    3.1  1.000000   
21  1969   6.68   5.5  19.700001  19.400000 -0.300001   5.34    4.2  2.900000   
22  1970   6.46   5.7  19.000000  19.299999  0.299999   6.68    5.5 -0.300001   
23  1971   4.35   4.4  17.299999  19.500000  2.200001   6.46    5.7  0.299999   
24  1972   4.07   3.2  17.600000  19.600000  2.000000   4.35    4.4  2.200001   
25  1973   7.04   6.2  17.600000  18.700001  1.100000   4.07    3.2  2.000000   
26  1974   7.89  11.0  18.299999  18.700001  0.400002   7.04    6.2  1.100000   
27  1975   5.84   9.1  17.900000  21.299999  3.400000   7.89   11.0  0.400002   
28  1976   4.99   5.8  17.100000  21.400000  4.299999   5.84    9.1  3.400000   
29  1977   5.27   6.5  18.000000  20.700001  2.700001   4.99    5.8  4.299999   
30  1978   7.22   7.6  18.000000  20.700001  2.700001   5.27    6.5  2.700001   
31  1979  10.04  11.3  18.500000  20.100000  1.600000   7.22    7.6  2.700001   
32  1980  11.51  13.5  19.000000  21.700001  2.700001  10.04   11.3  1.600000   
33  1981  14.03  10.3  19.600000  22.200001  2.600000  11.51   13.5  2.700001   
34  1982  10.69   6.2  19.200001  23.100000  3.900000  14.03   10.3  2.600000   
35  1983   8.63   3.2  17.400000  23.500000  6.100000  10.69    6.2  3.900000   
36  1984   9.58   4.3  17.299999  22.100000  4.800001   8.63    3.2  6.100000   
37  1985   7.48   3.6  17.700001  22.799999  5.099998   9.58    4.3  4.800001   
38  1986   5.98   1.9  17.500000  22.500000  5.000000   7.48    3.6  5.099998   
39  1987   5.82   3.6  18.400000  21.600000  3.200001   5.98    1.9  5.000000   
40  1988   6.69   4.1  18.100000  21.200001  3.100000   5.82    3.6  3.200001   
41  1989   8.12   4.8  18.299999  21.200001  2.900002   6.69    4.1  3.100000   
42  1990   7.51   5.4  18.000000  21.799999  3.799999   8.12    4.8  2.900002   
43  1991   5.42   4.2  17.799999  22.299999  4.500000   7.51    5.4  3.799999   
44  1992   3.45   3.0  17.500000  22.100000  4.600000   5.42    4.2  4.500000   
45  1993   3.02   3.0  17.500000  21.400000  3.900000   3.45    3.0  4.600000   
46  1994   4.29   2.6  18.100000  21.000000  2.900000   3.02    3.0  3.900000   
47  1995   5.51   2.8  18.500000  20.700001  2.200001   4.29    2.6  2.900000   
48  1996   5.02   3.0  18.900000  2

In [6]:
irate_regOLS = smf.ols(formula='i3 ~ Q("inf") + Q("def")', data = intdef)
results_irate_regOLS = irate_regOLS.fit()
print(results_irate_regOLS.summary())

                            OLS Regression Results                            
Dep. Variable:                     i3   R-squared:                       0.602
Model:                            OLS   Adj. R-squared:                  0.587
Method:                 Least Squares   F-statistic:                     40.09
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           2.48e-11
Time:                        15:36:04   Log-Likelihood:                -112.16
No. Observations:                  56   AIC:                             230.3
Df Residuals:                      53   BIC:                             236.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.7333      0.432      4.012      0.0

# Example 10.3: Puerto Rican Employment and the Minimum Wage

In [7]:
prminwge = woo.data('prminwge')
prminwge

year  avgmin  avgwage  kaitz  avgcov  covt  mfgwage  prdef  prepop  \
0   1950   0.198    0.398  0.155   0.201  0.29     0.43  0.859   0.470   
1   1951   0.209    0.410  0.164   0.207  0.29     0.45  0.881   0.449   
2   1952   0.225    0.421  0.180   0.226  0.29     0.48  0.953   0.434   
3   1953   0.311    0.480  0.229   0.231  0.29     0.50  0.970   0.428   
4   1954   0.313    0.508  0.211   0.224  0.29     0.52  1.000   0.415   
5   1955   0.369    0.547  0.231   0.236  0.29     0.57  1.003   0.419   
6   1956   0.447    0.601  0.257   0.245  0.29     0.64  1.011   0.412   
7   1957   0.488    0.685  0.251   0.244  0.29     0.76  1.035   0.412   
8   1958   0.555    0.716  0.258   0.238  0.29     0.83  1.089   0.397   
9   1959   0.588    0.789  0.266   0.260  0.29     0.87  1.110   0.394   
10  1960   0.616    0.840  0.268   0.270  0.29     0.92  1.138   0.403   
11  1961   0.608    0.875  0.251   0.269  0.29     0.99  1.173   0.397   
12  1962   0.707    0.933  0.270   0.279  0.29     1.06  1.216   0.385   
13  1963   0.723    1.036  0.255   0.279  0.29     1.13  1.247   0.395   
14  1964   0.809    1.097  0.274   0.294  0.31     1.18  1.298   0.396   
15  1965   0.834    1.176  0.271   0.302  0.31     1.23  1.327   0.401   
16  1966   0.854    1.288  0.325   0.444  0.44     1.29  1.358   0.399   
17  1967   0.971    1.371  0.365   0.448  0.44     1.39  1.421   0.399   
18  1968   1.104    1.512  0.381   0.455  0.44     1.55  1.500   0.403   
19  1969   1.149    1.667  0.360   0.455  0.49     1.65  1.552   0.399   
20  1970   1.209    1.856  0.347   0.458  0.47     1.76  1.616   0.428   
21  1971   1.224    1.990  0.331   0.457  0.47     1.87  1.708   0.423   
22  1972   1.257    2.144  0.316   0.449  0.47     2.00  1.780   0.423   
23  1973   1.262    2.281  0.304   0.452  0.47     2.13  1.817   0.421   
24  1974   1.681    2.452  0.381   0.544  0.60     2.32  1.946   0.405   
25  1975   1.871    2.562  0.434   0.594  0.66     2.56  2.082   0.368   
26  1976   2.034    2.681  0.442   0.588  0.64     2.78  2.174   0.364   
27  1977   2.198    3.023  0.435   0.590  0.64     3.02  2.240   0.358   
28  1978   2.509    3.323  0.456   0.596  0.64     3.36  2.340   0.362   
29  1979   2.768    3.589  0.468   0.594  0.64     3.69  2.483   0.360   
30  1980   2.997    3.883  0.461   0.589  0.64     4.02  2.716   0.359   
31  1981   3.264    4.181  0.467   0.587  0.64     4.39  2.954   0.343   
32  1982   3.305    4.318  0.461   0.585  0.64     4.64  3.175   0.318   
33  1983   3.350    4.456  0.454   0.583  0.64     4.83  3.321   0.321   
34  1984   3.350    4.498  0.449   0.586  0.64     5.02  3.461   0.334   
35  1985   3.350    4.565  0.440   0.579  0.64     5.19  3.548   0.331   
36  1986   3.350    4.725  0.426   0.581  0.64     5.31  3.697   0.351   
37  1987   3.350    4.879  0.409   0.582  0.64     5.33  3.787   0.369   

    prepopf  ...  lprunemp    lprgnp    lusgnp    lkaitz   lprun_1   lprepop  \
0     0.470  ...  2.734367  6.778443  7.093155 -1.864330       NaN -0.755023   
1     0.449  ...  2.772589  6.829794  7.191580 -1.807889 -1.870803 -0.800732   
2     0.434  ...  2.694627  6.923530  7.229839 -1.714798 -1.832582 -0.834711   
3     0.428  ...  2.674149  6.985919  7.269129 -1.474033 -1.910543 -0.848632   
4     0.415  ...  2.727853  7.007058  7.255733 -1.555897 -1.931022 -0.879477   
5     0.419  ...  2.580217  7.037467  7.309814 -1.465338 -1.877317 -0.869884   
6     0.412  ...  2.587764  7.077582  7.330143 -1.358679 -2.024953 -0.886732   
7     0.412  ...  2.549445  7.108080  7.346720 -1.382302 -2.017406 -0.886732   
8     0.397  ...  2.653242  7.137597  7.339018 -1.354796 -2.055725 -0.923819   
9     0.394  ...  2.587764  7.217884  7.395783 -1.324259 -1.951928 -0.931404   
10    0.403  ...  2.468100  7.295192  7.417760 -1.316768 -2.017406 -0.908819   
11    0.397  ...  2.541602  7.354234  7.443488 -1.382302 -2.137071 -0.923819   
12    0.385  ...  2.549445  7.428868  7.495209 -1.309333 -2.063568 -0.954512 

In [8]:
prminwge_ols = smf.ols('lprepop ~ lmincov + lusgnp', data = prminwge)
results_prminwge_ols = prminwge_ols.fit()
print(results_prminwge_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                lprepop   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.641
Method:                 Least Squares   F-statistic:                     34.04
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           6.17e-09
Time:                        15:36:04   Log-Likelihood:                 57.376
No. Observations:                  38   AIC:                            -108.8
Df Residuals:                      35   BIC:                            -103.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.0544      0.765     -1.378      0.1

# Example 10.4: Effects of Personal Exemption on Fertility Rates

In [9]:
fertil3 = woo.data('fertil3')
fertil3

gfr          pe  year   t   tsq        pe_1        pe_2  \
0   124.699997    0.000000  1913   1     1         NaN         NaN   
1   126.599998    0.000000  1914   2     4    0.000000         NaN   
2   125.000000    0.000000  1915   3     9    0.000000    0.000000   
3   123.400002    0.000000  1916   4    16    0.000000    0.000000   
4   121.000000   19.270000  1917   5    25    0.000000    0.000000   
..         ...         ...   ...  ..   ...         ...         ...   
67   68.400002  123.169998  1980  68  4624  132.929993  118.980003   
68   67.400002  119.309998  1981  69  4761  123.169998  132.929993   
69   67.300003  102.040001  1982  70  4900  119.309998  123.169998   
70   65.800003   92.489998  1983  71  5041  102.040001  119.309998   
71   65.400002   83.900002  1984  72  5184   92.489998  102.040001   

          pe_3        pe_4  pill  ...      cpe_1      cpe_2      cpe_3  \
0          NaN         NaN     0  ...        NaN        NaN        NaN   
1          NaN         NaN     0  ...        NaN        NaN        NaN   
2          NaN         NaN     0  ...   0.000000        NaN        NaN   
3     0.000000         NaN     0  ...   0.000000   0.000000        NaN   
4     0.000000    0.000000     0  ...   0.000000   0.000000   0.000000   
..         ...         ...   ...  ...        ...        ...        ...   
67  116.110001  120.080002     1  ...  13.949989   2.870003  -3.970001   
68  118.980003  116.110001     1  ...  -9.759995  13.949989   2.870003   
69  132.929993  118.980003     1  ...  -3.860001  -9.759995  13.949989   
70  123.169998  132.929993     1  ... -17.269997  -3.860001  -9.759995   
71  119.309998  123.169998     1  ...  -9.550003 -17.269997  -3.860001   

        cpe_4       gfr_1    cgfr_1    cgfr_2    cgfr_3    cgfr_4       gfr_2  
0         NaN         NaN       NaN       NaN       NaN       NaN         NaN  
1         NaN  124.699997       NaN       NaN       NaN       NaN         NaN  
2         NaN  126.599998  1.900002       NaN       NaN       NaN  124.699997  
3         NaN  125.000000 -1.599998  1.900002       NaN       NaN  126.599998  
4         NaN  123.400002 -1.599998 -1.599998  1.900002       NaN  125.000000  
..        ...         ...       ...       ...       ...       ...         ...  
67  -2.279999   67.199997  1.699997 -1.300003  1.000000 -0.199997   65.500000  
68  -3.970001   68.400002  1.200005  1.699997 -1.300003  1.000000   67.199997  
69   2.870003   67.400002 -1.000000  1.200005  1.699997 -1.300003   68.400002  
70  13.949989   67.300003 -0.099998 -1.000000  1.200005  1.699997   67.400002  
71  -9.759995   65.800003 -1.500000 -0.099998 -1.000000  1.200005   67.300003  

[72 rows x 24 columns]

In [10]:
T = len(fertil3)

In [11]:
# define yearly time series beginning in 1913:

fertil3.index = pd.date_range(start = '1913', periods = T, freq = 'Y').year

In [12]:
static_fertil_ols = smf.ols('gfr ~ pe + ww2 + pill', data = fertil3)
results_static_fertil_ols = static_fertil_ols.fit()
print(results_static_fertil_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                    gfr   R-squared:                       0.473
Model:                            OLS   Adj. R-squared:                  0.450
Method:                 Least Squares   F-statistic:                     20.38
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           1.57e-09
Time:                        15:36:05   Log-Likelihood:                -293.56
No. Observations:                  72   AIC:                             595.1
Df Residuals:                      68   BIC:                             604.2
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     98.6818      3.208     30.760      0.0

In [13]:
# fdl = finite distributed lags model

fdl_fertil_ols = smf.ols('gfr ~ pe  + pe_1 + pe_2 + ww2 + pill', data = fertil3)
results_fdl_fertil_ols = fdl_fertil_ols.fit()
print(results_fdl_fertil_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                    gfr   R-squared:                       0.499
Model:                            OLS   Adj. R-squared:                  0.459
Method:                 Least Squares   F-statistic:                     12.73
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           1.35e-08
Time:                        15:36:05   Log-Likelihood:                -282.26
No. Observations:                  70   AIC:                             576.5
Df Residuals:                      64   BIC:                             590.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     95.8705      3.282     29.211      0.0

# Example 10.5: Antidumping Filings and Chemical Imports

In [14]:
barium = woo.data('barium')
T = len(barium)
barium.index = pd.date_range(start = '1978', periods = T, freq = 'M')
barium

chnimp       bchlimp  befile6  affile6  afdec6  befile12  \
1978-01-31   220.462006   9578.375977        0        0       0         0   
1978-02-28    94.797997  11219.480469        0        0       0         0   
1978-03-31   219.357498   9719.900391        0        0       0         0   
1978-04-30   317.421509  12920.950195        0        0       0         0   
1978-05-31   114.639000   9790.446289        0        0       0         0   
...                 ...           ...      ...      ...     ...       ...   
1988-07-31   589.421997  31686.859375        0        0       0         0   
1988-08-31   529.733521  16370.339844        0        0       0         0   
1988-09-30   968.037476  15653.959961        0        0       0         0   
1988-10-31   794.059998  45716.609375        0        0       0         0   
1988-11-30  1087.806030  29580.500000        0        0       0         0   

            affile12  afdec12      chempi           gas  ...  apr  may  jun  \
1978-01-31         0        0  100.099998  7.830000e+09  ...    0    0    0   
1978-02-28         0        0  100.900002  8.820000e+09  ...    0    0    0   
1978-03-31         0        0  101.099998  8.450000e+09  ...    1    0    0   
1978-04-30         0        0  102.500000  9.240001e+09  ...    0    1    0   
1978-05-31         0        0  104.099998  9.150000e+09  ...    0    0    1   
...              ...      ...         ...           ...  ...  ...  ...  ...   
1988-07-31         0        0  154.800003  9.420000e+09  ...    0    0    0   
1988-08-31         0        0  155.300003  8.790000e+09  ...    0    0    0   
1988-09-30         0        0  156.699997  8.990000e+09  ...    0    0    0   
1988-10-31         0        0  157.500000  8.920001e+09  ...    0    0    0   
1988-11-30         0        0  159.000000  9.540000e+09  ...    0    0    0   

            jul  aug  sep  oct  nov  dec   percchn  
1978-01-31    0    0    0    0    0    0  2.301664  
1978-02-28    0    0    0    0    0    0  0.844941  
1978-03-31    0    0    0    0    0    0  2.256788  
1978-04-30    0    0    0    0    0    0  2.456642  
1978-05-31    0    0    0    0    0    0  1.170927  
...         ...  ...  ...  ...  ...  ...       ...  
1988-07-31    0    1    0    0    0    0  1.860147  
1988-08-31    0    0    1    0    0    0  3.235935  
1988-09-30    0    0    0    1    0    0  6.183978  
1988-10-31    0    0    0    0    1    0  1.736918  
1988-11-30    0    0    0    0    0    1  3.677443  

[131 rows x 31 columns]

In [15]:
barium_ols = smf.ols('lchnimp ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6', data = barium)
results_barium_ols = barium_ols.fit()
print(results_barium_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                lchnimp   R-squared:                       0.305
Model:                            OLS   Adj. R-squared:                  0.271
Method:                 Least Squares   F-statistic:                     9.064
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           3.25e-08
Time:                        15:36:06   Log-Likelihood:                -114.79
No. Observations:                 131   AIC:                             243.6
Df Residuals:                     124   BIC:                             263.7
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -17.8030     21.045     -0.846      0.3

# Example 10.6: Election Outcomes and Economic Performance

In [16]:
fair = woo.data('fair')
fair

year       V  I  DPER   DUR      g3     p15   n         g2     gYR  ...  \
0   1916  0.5168  1     1  0.00   2.229   4.252   3  -1.213000   6.035  ...   
1   1920  0.3612  1     0  1.00 -11.463  16.535   5 -14.496000  -8.147  ...   
2   1924  0.4176 -1    -1  0.00  -3.872   5.161  10  -9.777000  -1.009  ...   
3   1928  0.4118 -1     0 -1.00   4.623   0.183   7   6.043000  -0.225  ...   
4   1932  0.5916 -1    -1 -1.25 -15.574   6.657   3 -16.249001 -14.369  ...   
5   1936  0.6246  1     1  0.00  12.625   3.387   9  18.764999  12.933  ...   
6   1940  0.5500  1     1  1.00   2.420   0.553   8  10.350000   6.708  ...   
7   1944  0.5377  1     1  1.25   2.910   6.432  13   3.317000   5.768  ...   
8   1948  0.5237  1     1  1.50   3.105  10.369   3   3.976000   2.289  ...   
9   1952  0.4460  1     0  1.75   0.910   2.256   7   0.632000   1.935  ...   
10  1956  0.4224 -1    -1  0.00  -1.479   2.132   6  -0.576000   0.338  ...   
11  1960  0.5009 -1     0 -1.00   0.020   2.299   5  -2.940000  -0.038  ...   
12  1964  0.6134  1     1  0.00   4.950   1.201  11   3.334000   4.374  ...   
13  1968  0.4960  1     0  1.00   4.712   3.160   9   4.125000   3.599  ...   
14  1972  0.3821 -1    -1  0.00   5.716   4.762   6   5.216000   4.247  ...   
15  1976  0.5105 -1     0 -1.00   3.411   7.604   6   0.923000   4.406  ...   
16  1980  0.4470  1     1  0.00  -3.512   7.947   5  -5.494000  -1.330  ...   
17  1984  0.4083 -1    -1  0.00   5.722   5.296   7   3.965000   6.114  ...   
18  1988  0.4610 -1     0 -1.00   2.174   3.392   5   2.285000   2.943  ...   
19  1992  0.5345 -1    -1 -1.25   1.478   3.834   0   1.102000   0.979  ...   
20  1996  0.5465  1     1  0.00   2.100   3.019   3   2.200000   1.500  ...   

       Ig3  Ip151md  In1md  demvote  incum  gnews     inf  partyWH  pWHgnews  \
0    2.229    4.252      3   0.5168      1      3   4.252        1         3   
1  -11.463    0.000      0   0.3612      0      5  16.535        1         5   
2    3.872   -5.161    -10   0.4176     -1     10   5.161       -1       -10   
3   -4.623   -0.183     -7   0.4118      0      7   0.183       -1        -7   
4   15.574   -6.657     -3   0.5916     -1      3   6.657       -1        -3   
5   12.625    3.387      9   0.6246      1      9   3.387        1         9   
6    2.420    0.553      8   0.5500      1      8   0.553        1         8   
7    2.910    0.000      0   0.5377      1     13   6.432        1        13   
8    3.105    0.000      0   0.5237      1      3  10.369        1         3   
9    0.910    2.256      7   0.4460      0      7   2.256        1         7   
10   1.479   -2.132     -6   0.4224     -1      6   2.132       -1        -6   
11  -0.020   -2.299     -5   0.5009      0      5   2.299       -1        -5   
12   4.950    1.201     11   0.6134      1     11   1.201        1        11   
13   4.712    3.160      9   0.4960      0      9   3.160        1         9   
14  -5.716   -4.762     -6   0.3821     -1      6   4.762       -1        -6   
15  -3.411   -7.604     -6   0.5105      0      6   7.604       -1        -6   
16  -3.512    7.947      5   0.4470      1      5   7.947        1         5   
17  -5.722   -5.296     -7   0.4083     -1      7   5.296       -1        -7   
18  -2.174   -3.392     -5   0.4610      0      5   3.392       -1        -5   
19  -1.478   -3.834      0   0.5345     -1      0   3.834       -1         0   
20   2.100    3.000      2   0.5465      1      3   3.019        1         3   

    pWHinf  
0    4.252  
1   16.535  
2   -5.161  
3   -0.183  
4   -6.657  
5    3.387  
6    0.553  
7    6.432  
8   10.369  
9    2.256  
10  -2.132  
11  -2.299  
12   1.201  
13   3.160  
14  -4.762  
15  -7.604  
16   7.947  
17  -5.296  
18  -3.392  
19  -3.834  
20   3.019  

[21 rows x 28 columns]

In [17]:
election_ols = smf.ols('demvote ~ partyWH + incum + pWHgnews + pWHinf', data = fair)
results_election_ols = election_ols.fit()
print(results_election_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                demvote   R-squared:                       0.659
Model:                            OLS   Adj. R-squared:                  0.574
Method:                 Least Squares   F-statistic:                     7.725
Date:                Wed, 27 Jul 2022   Prob (F-statistic):            0.00115
Time:                        15:36:06   Log-Likelihood:                 36.393
No. Observations:                  21   AIC:                            -62.79
Df Residuals:                      16   BIC:                            -57.56
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4842      0.011     42.210      0.0

# Example 10.7: Housing Investment and Prices

In [18]:
hseinv = woo.data('hseinv')
hseinv

year       inv       pop   price       linv       lpop    lprice   t  \
0   1947   54864.0  144126.0  0.8190  10.912613  11.878443 -0.199671   1   
1   1948   64717.0  146631.0  0.8649  11.077779  11.895675 -0.145141   2   
2   1949   63150.0  149188.0  0.8456  11.053268  11.912963 -0.167709   3   
3   1950   86014.0  151684.0  0.8765  11.362266  11.929555 -0.131819   4   
4   1951   70610.0  154287.0  0.8819  11.164927  11.946569 -0.125677   5   
5   1952   68574.0  156954.0  0.8842  11.135669  11.963708 -0.123072   6   
6   1953   70818.0  159565.0  0.8868  11.167869  11.980206 -0.120136   7   
7   1954   78460.0  162391.0  0.8597  11.270344  11.997763 -0.151172   8   
8   1955   91204.0  165275.0  0.8708  11.420854  12.015366 -0.138343   9   
9   1956   80383.0  168221.0  0.8829  11.294558  12.033034 -0.124543  10   
10  1957   74040.0  171274.0  0.8722  11.212360  12.051020 -0.136737  11   
11  1958   74822.0  174141.0  0.8521  11.222867  12.067620 -0.160051  12   
12  1959   88936.0  177830.0  0.8647  11.395672  12.088583 -0.145373  13   
13  1960   83127.0  180671.0  0.8620  11.328125  12.104433 -0.148500  14   
14  1961   83207.0  183691.0  0.8553  11.329086  12.121010 -0.156303  15   
15  1962   89121.0  186538.0  0.8593  11.397750  12.136391 -0.151637  16   
16  1963   96778.0  189242.0  0.8656  11.480175  12.150782 -0.144332  17   
17  1964   94306.0  191889.0  0.8795  11.454300  12.164672 -0.128402  18   
18  1965  100103.0  194303.0  0.8774  11.513955  12.177174 -0.130792  19   
19  1966   92145.0  196560.0  0.8783  11.431119  12.188723 -0.129767  20   
20  1967   91336.0  198712.0  0.8919  11.422300  12.199612 -0.114401  21   
21  1968   99617.0  200706.0  0.9036  11.509089  12.209597 -0.101369  22   
22  1969  102183.0  202677.0  0.9152  11.534520  12.219369 -0.088613  23   
23  1970   96855.0  205052.0  0.8823  11.480970  12.231019 -0.125223  24   
24  1971  124309.0  207191.0  0.8798  11.730526  12.241396 -0.128061  25   
25  1972  147269.0  209353.0  0.8874  11.900016  12.251777 -0.119459  26   
26  1973  145286.0  211536.0  0.9057  11.886459  12.262151 -0.099047  27   
27  1974  112349.0  213743.0  0.9232  11.629365  12.272530 -0.079909  28   
28  1975   95170.0  215973.0  0.9147  11.463420  12.282908 -0.089159  29   
29  1976  117727.0  218280.0  0.9199  11.676124  12.293534 -0.083490  30   
30  1977  142412.0  220612.0  0.9604  11.866480  12.304161 -0.040405  31   
31  1978  151261.0  222968.0  1.0061  11.926762  12.314783  0.006082  32   
32  1979  143049.0  225350.0  1.0356  11.870942  12.325410  0.034981  33   
33  1980  112310.0  227757.0  1.0432  11.629018  12.336035  0.042293  34   
34  1981  102714.0  230139.0  1.0282  11.539703  12.346438  0.027810  35   
35  1982   84676.0  232520.0  1.0000  11.346587  12.356731  0.000000  36   
36  1983  122819.0  234799.0  0.9836  11.718467  12.366485 -0.016536  37   
37  1984  145166.0  237001.0  0.9836  11.885633  12.375819 -0.016536  38   
38  1985  146311.0  239279.0  0.9760  11.893490  12.385386 -0.024293  39   
39  1986  168406.0  241625.0  0.9760  12.034133  12.395143 -0.024293  40   
40  1987  167459.0  243934.0  0.9892  12.028494  12.404653 -0.010859  41   
41  1988  165459.0  246329.0  0.9864  12.016479  12.414423 -0.013693  42   

       invpc    linvpc  lprice_1  linvpc_1    gprice    ginvpc  
0   0.380667 -0.965831       NaN       NaN       NaN       NaN  
1   0.441360 -0.817895 -0.199671 -0.965831  0.054530  0.147935  
2   0.423291 -0.859694 -0.145141 -0.817895 -0.022567 -0.041799  
3   0.567060 -0.567289 -0.167709 -0.859694  0.035890  0.292405  
4   0.457654 -0.781643 -0.131819 -0.567289  0.006142 -0.214353  
5   0.436905 -0.828039 -0.125677 -0.781643  0.002605 -0.046397  
6   0.443819 -0.812338 -0.123072 -0.828039  0.002936  0.015701  
7   0.483155 -0.727418 -0.120136 -0.812338 -0.031036  0.084920  
8   0.551832 -0.594512 -0.151172 -0.727418  0.012829  0.132906  
9   0.477842 -0.738476 -0.138343 -0.594512  0.013800 -0.143964  
10  0.432290 -0.8386

In [19]:
# linear regression without time trend:

reg_hseinv_wot = smf.ols('linvpc ~ lprice', data = hseinv)
results_reg_hseinv_wot = reg_hseinv_wot.fit()
print(results_reg_hseinv_wot.summary())

                            OLS Regression Results                            
Dep. Variable:                 linvpc   R-squared:                       0.208
Model:                            OLS   Adj. R-squared:                  0.189
Method:                 Least Squares   F-statistic:                     10.53
Date:                Wed, 27 Jul 2022   Prob (F-statistic):            0.00238
Time:                        15:36:06   Log-Likelihood:                 19.617
No. Observations:                  42   AIC:                            -35.23
Df Residuals:                      40   BIC:                            -31.76
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.5502      0.043    -12.788      0.0

In [20]:
# linear regression with time trend:

reg_hseinv_time = smf.ols('linvpc ~ lprice + t', data = hseinv)
results_reg_hseinv_time = reg_hseinv_time.fit()
print(results_reg_hseinv_time.summary())

                            OLS Regression Results                            
Dep. Variable:                 linvpc   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.307
Method:                 Least Squares   F-statistic:                     10.08
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           0.000296
Time:                        15:36:07   Log-Likelihood:                 23.459
No. Observations:                  42   AIC:                            -40.92
Df Residuals:                      39   BIC:                            -35.71
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.9131      0.136     -6.733      0.0

# Example 10.8: Fertility Equation

In [21]:
fertil3

gfr          pe  year   t   tsq        pe_1        pe_2  \
1913  124.699997    0.000000  1913   1     1         NaN         NaN   
1914  126.599998    0.000000  1914   2     4    0.000000         NaN   
1915  125.000000    0.000000  1915   3     9    0.000000    0.000000   
1916  123.400002    0.000000  1916   4    16    0.000000    0.000000   
1917  121.000000   19.270000  1917   5    25    0.000000    0.000000   
...          ...         ...   ...  ..   ...         ...         ...   
1980   68.400002  123.169998  1980  68  4624  132.929993  118.980003   
1981   67.400002  119.309998  1981  69  4761  123.169998  132.929993   
1982   67.300003  102.040001  1982  70  4900  119.309998  123.169998   
1983   65.800003   92.489998  1983  71  5041  102.040001  119.309998   
1984   65.400002   83.900002  1984  72  5184   92.489998  102.040001   

            pe_3        pe_4  pill  ...      cpe_1      cpe_2      cpe_3  \
1913         NaN         NaN     0  ...        NaN        NaN        NaN   
1914         NaN         NaN     0  ...        NaN        NaN        NaN   
1915         NaN         NaN     0  ...   0.000000        NaN        NaN   
1916    0.000000         NaN     0  ...   0.000000   0.000000        NaN   
1917    0.000000    0.000000     0  ...   0.000000   0.000000   0.000000   
...          ...         ...   ...  ...        ...        ...        ...   
1980  116.110001  120.080002     1  ...  13.949989   2.870003  -3.970001   
1981  118.980003  116.110001     1  ...  -9.759995  13.949989   2.870003   
1982  132.929993  118.980003     1  ...  -3.860001  -9.759995  13.949989   
1983  123.169998  132.929993     1  ... -17.269997  -3.860001  -9.759995   
1984  119.309998  123.169998     1  ...  -9.550003 -17.269997  -3.860001   

          cpe_4       gfr_1    cgfr_1    cgfr_2    cgfr_3    cgfr_4  \
1913        NaN         NaN       NaN       NaN       NaN       NaN   
1914        NaN  124.699997       NaN       NaN       NaN       NaN   
1915        NaN  126.599998  1.900002       NaN       NaN       NaN   
1916        NaN  125.000000 -1.599998  1.900002       NaN       NaN   
1917        NaN  123.400002 -1.599998 -1.599998  1.900002       NaN   
...         ...         ...       ...       ...       ...       ...   
1980  -2.279999   67.199997  1.699997 -1.300003  1.000000 -0.199997   
1981  -3.970001   68.400002  1.200005  1.699997 -1.300003  1.000000   
1982   2.870003   67.400002 -1.000000  1.200005  1.699997 -1.300003   
1983  13.949989   67.300003 -0.099998 -1.000000  1.200005  1.699997   
1984  -9.759995   65.800003 -1.500000 -0.099998 -1.000000  1.200005   

           gfr_2  
1913         NaN  
1914         NaN  
1915  124.699997  
1916  126.599998  
1917  125.000000  
...          ...  
1980   65.500000  
1981   67.199997  
1982   68.400002  
1983   67.400002  
1984   67.300003  

[72 rows x 24 columns]

In [23]:
# Adding time trends on fertility equation

# Regular time trend

fertil_ols_t = smf.ols('gfr ~ pe + ww2 + pill + t', data = fertil3)
results_fertil_ols_t = fertil_ols_t.fit()

# Regular and squared time trend

fertil_ols_tsq = smf.ols('gfr ~ pe + ww2 + pill + t + tsq', data = fertil3)
results_fertil_ols_tsq = fertil_ols_tsq.fit()

print(results_fertil_ols_t.summary(), results_fertil_ols_tsq.summary())

                            OLS Regression Results                            
Dep. Variable:                    gfr   R-squared:                       0.662
Model:                            OLS   Adj. R-squared:                  0.642
Method:                 Least Squares   F-statistic:                     32.84
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           3.76e-15
Time:                        15:49:46   Log-Likelihood:                -277.57
No. Observations:                  72   AIC:                             565.1
Df Residuals:                      67   BIC:                             576.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    111.7694      3.358     33.287      0.0

# Example 10.9: Puerto Rican Employment

In [24]:
prminwge_ols_t = smf.ols('lprepop ~ lmincov + lusgnp + t', data = prminwge)
results_prminwge_ols_t = prminwge_ols_t.fit()
print(results_prminwge_ols_t.summary())

                            OLS Regression Results                            
Dep. Variable:                lprepop   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.834
Method:                 Least Squares   F-statistic:                     62.78
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           6.01e-14
Time:                        15:59:30   Log-Likelihood:                 72.532
No. Observations:                  38   AIC:                            -137.1
Df Residuals:                      34   BIC:                            -130.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.6963      1.296     -6.711      0.0

# Example 10.11: Effects of Antidumping Filings

In [31]:
barium_ols_season = smf.ols('lchnimp ~ lchempi + lgas + lrtwex + befile6 + affile6 + afdec6 + mar + apr + may + jun + jul + aug + sep + oct + nov + dec', 
                            data = barium)
results_barium_ols_season = barium_ols_season.fit()
print(results_barium_ols_season.summary())

                            OLS Regression Results                            
Dep. Variable:                lchnimp   R-squared:                       0.348
Model:                            OLS   Adj. R-squared:                  0.256
Method:                 Least Squares   F-statistic:                     3.797
Date:                Wed, 27 Jul 2022   Prob (F-statistic):           1.30e-05
Time:                        16:20:06   Log-Likelihood:                -110.63
No. Observations:                 131   AIC:                             255.3
Df Residuals:                     114   BIC:                             304.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -5.8255     28.040     -0.208      0.8